Python Linting and Formatting with ruff (python_lint, python_format)

## Priority: P1 - High Priority

**Part of Epic:** #131  
**Milestone:** 2025-Q1 Phase 1  
**Estimated Effort:** 10-14 hours

## Problem Statement

Python developers need unified linting and formatting capabilities using ruff, the modern all-in-one tool that replaces flake8, black, isort, pyupgrade, and more. This provides a consistent, fast development experience following 2025 Python best practices.

**Why ruff over traditional tools:**
- **10-100x faster** than flake8/black/isort combined (written in Rust)
- **Single tool** replaces 4-5 separate tools
- **Better defaults** aligned with modern Python practices
- **Auto-fix** capabilities for most issues
- **Growing ecosystem** adoption (used by major projects)

## Proposed Solution

Implement two tools: `python_lint` and `python_format` following the patterns from `src/tools/lint-tools.ts` and `go_fmt`.

**Tool Signatures:**
```typescript
async pythonLint(args: PythonLintArgs): Promise<PythonToolResult>
async pythonFormat(args: PythonFormatArgs): Promise<PythonToolResult>
```

**python_lint Features:**
- Run ruff check with configurable rule sets
- Support auto-fix mode
- Show detailed error explanations
- Output in multiple formats (text, JSON, GitHub)
- Respect pyproject.toml configuration

**python_format Features:**
- Format code with ruff format
- Check mode (verify without modifying)
- Respect line length and style settings
- Preview unstable formatting features

## Implementation Guidance

### Development Steps

1. **Study Existing Lint Tools**
   ```bash
   cat src/tools/lint-tools.ts | grep -A 50 "markdownlint"
   cat src/tools/go-tools.ts | grep -A 50 "goFmt"
   ```

2. **Define Zod Schemas**
   ```typescript
   const PythonLintSchema = z.object({
     directory: z.string().optional(),
     files: z.array(z.string()).optional().describe('Specific files to lint'),
     fix: z.boolean().optional().default(false).describe('Auto-fix issues'),
     select: z.array(z.string()).optional().describe('Rule codes to enable (e.g., ["E", "F", "I"])'),
     ignore: z.array(z.string()).optional().describe('Rule codes to ignore'),
     outputFormat: z.enum(['text', 'json', 'github']).optional().default('text'),
     showFixes: z.boolean().optional().default(false),
   });
   
   const PythonFormatSchema = z.object({
     directory: z.string().optional(),
     files: z.array(z.string()).optional().describe('Specific files to format'),
     check: z.boolean().optional().default(false).describe('Check without modifying'),
     lineLength: z.number().optional().describe('Max line length'),
     preview: z.boolean().optional().default(false).describe('Enable preview style'),
   });
   ```

3. **Implement python_lint**
   ```typescript
   async pythonLint(args: PythonLintArgs): Promise<PythonToolResult> {
     const validated = PythonLintSchema.parse(args);
     const directory = validated.directory || process.cwd();
     
     // Build ruff check command
     const cmd = ['ruff', 'check'];
     
     // Add fix mode
     if (validated.fix) {
       cmd.push('--fix');
     }
     
     // Add output format
     if (validated.outputFormat !== 'text') {
       cmd.push('--output-format', validated.outputFormat);
     }
     
     // Add rule selection
     if (validated.select?.length) {
       cmd.push('--select', validated.select.join(','));
     }
     
     // Add rule ignores
     if (validated.ignore?.length) {
       cmd.push('--ignore', validated.ignore.join(','));
     }
     
     // Add show fixes
     if (validated.showFixes) {
       cmd.push('--show-fixes');
     }
     
     // Add files or use current directory
     if (validated.files?.length) {
       cmd.push(...validated.files);
     } else {
       cmd.push('.');
     }
     
     const result = await this.shellExecutor.executeCommand(
       cmd.join(' '),
       directory,
       { timeout: 60000 } // 1 minute
     );
     
     return this.parseRuffLintOutput(result, validated.outputFormat);
   }
   ```

4. **Implement python_format**
   ```typescript
   async pythonFormat(args: PythonFormatArgs): Promise<PythonToolResult> {
     const validated = PythonFormatSchema.parse(args);
     const directory = validated.directory || process.cwd();
     
     // Build ruff format command
     const cmd = ['ruff', 'format'];
     
     // Add check mode
     if (validated.check) {
       cmd.push('--check');
     }
     
     // Add line length
     if (validated.lineLength) {
       cmd.push('--line-length', validated.lineLength.toString());
     }
     
     // Add preview mode
     if (validated.preview) {
       cmd.push('--preview');
     }
     
     // Add files or use current directory
     if (validated.files?.length) {
       cmd.push(...validated.files);
     } else {
       cmd.push('.');
     }
     
     const result = await this.shellExecutor.executeCommand(
       cmd.join(' '),
       directory,
       { timeout: 60000 }
     );
     
     return this.parseRuffFormatOutput(result);
   }
   ```

5. **Parse ruff Output**
   ```typescript
   private parseRuffLintOutput(
     result: ExecutionResult, 
     format: string
   ): PythonToolResult {
     if (format === 'json') {
       try {
         const issues = JSON.parse(result.stdout);
         return {
           success: issues.length === 0,
           data: { issues, count: issues.length },
         };
       } catch (e) {
         return { success: false, error: 'Failed to parse JSON output' };
       }
     }
     
     // Parse text output
     const issueCount = (result.stdout.match(/\n/g) || []).length;
     return {
       success: result.exitCode === 0,
       data: { 
         output: result.stdout,
         issuesFound: issueCount,
         fixed: result.stdout.includes('fixed'),
       },
     };
   }
   
   private parseRuffFormatOutput(result: ExecutionResult): PythonToolResult {
     const filesChanged = (result.stdout.match(/\d+ file/)?.[0] || '0 files').split(' ')[0];
     
     return {
       success: result.exitCode === 0,
       data: {
         filesFormatted: parseInt(filesChanged),
         output: result.stdout,
       },
     };
   }
   ```

6. **Update ShellExecutor Allowlist**
   ```typescript
   // Add to ALLOWED_COMMANDS
   'ruff',
   ```

7. **Add Error Handling**
   - Detect missing ruff: `command not found` → suggest `pip install ruff` or `uv tool install ruff`
   - Detect configuration errors → suggest checking pyproject.toml
   - Detect syntax errors → show helpful context

### Testing Requirements

Create comprehensive tests in `src/__tests__/tools/python-tools.test.ts`:

**Test Coverage (Target: 85-90%+):**
- [ ] Lints Python files successfully
- [ ] Auto-fixes issues when fix=true
- [ ] Respects select/ignore rules
- [ ] Outputs in different formats (text, JSON, GitHub)
- [ ] Formats Python files successfully
- [ ] Check mode doesn't modify files
- [ ] Respects line length settings
- [ ] Handles missing ruff gracefully
- [ ] Handles syntax errors in Python files
- [ ] Works with specific file lists
- [ ] Works in subdirectories
- [ ] Parses issue counts correctly

**Test Example:**
```typescript
describe('PythonTools - lint and format', () => {
  it('lints Python code and finds issues', async () => {
    mockShellExecutor.mockCommand('ruff check .', `
      src/main.py:10:1: F401 [*] \`os\` imported but unused
      src/main.py:25:80: E501 Line too long (89 > 88)
      Found 2 errors.
    `);
    
    const result = await pythonTools.pythonLint({});
    
    expect(result.success).toBe(false);
    expect(result.data.issuesFound).toBe(2);
  });
  
  it('auto-fixes issues when fix is enabled', async () => {
    mockShellExecutor.mockCommand('ruff check --fix .', `
      src/main.py:10:1: F401 [*] \`os\` imported but unused
      Fixed 1 error.
    `);
    
    const result = await pythonTools.pythonLint({ fix: true });
    
    expect(result.data.fixed).toBe(true);
  });
  
  it('formats Python code', async () => {
    mockShellExecutor.mockCommand('ruff format .', '3 files reformatted');
    
    const result = await pythonTools.pythonFormat({});
    
    expect(result.success).toBe(true);
    expect(result.data.filesFormatted).toBe(3);
  });
});
```

### Linting & Validation

**MUST RUN BEFORE PR:**
```bash
make lint       # TypeScript/JavaScript linting
make test       # Full test suite
make build      # Verify build passes
```

### Documentation Updates

1. **JSDoc Comments**
   ```typescript
   /**
    * Lint Python code using ruff
    * 
    * Ruff is a fast, all-in-one linter that replaces flake8, isort,
    * pyupgrade, and more. Written in Rust for 10-100x performance.
    * 
    * @param args - Lint configuration
    * @returns Lint results with issue count and details
    * 
    * @example
    * ```typescript
    * // Lint and auto-fix
    * await pythonTools.pythonLint({ fix: true });
    * 
    * // Lint with specific rules
    * await pythonTools.pythonLint({ 
    *   select: ['E', 'F', 'I'],
    *   outputFormat: 'json'
    * });
    * ```
    */
   ```

2. **Update README.md**
   - Add ruff documentation
   - Show common rule sets
   - Document auto-fix capabilities

## Acceptance Criteria

### Feature Complete
- [ ] `python_lint` tool implemented with ruff check
- [ ] `python_format` tool implemented with ruff format
- [ ] Auto-fix support working
- [ ] Multiple output formats (text, JSON, GitHub)
- [ ] Rule selection/ignoring working
- [ ] Error messages guide users to install ruff
- [ ] Code follows lint-tools pattern

### Quality Gates (MUST PASS)
- [ ] `make lint` passes with zero errors
- [ ] `make test` passes with 85-90%+ coverage
- [ ] `make build` completes successfully
- [ ] All tests are meaningful
- [ ] Error messages are actionable

### Documentation
- [ ] JSDoc comments added
- [ ] README.md updated with ruff examples
- [ ] Common rule sets documented

### PR Ready
- [ ] PR_MESSAGE.md generated
- [ ] Related to Epic #131
- [ ] All acceptance criteria met

## Example Use Cases

### Use Case 1: Lint with Auto-Fix
```bash
{
  "tool": "python_lint",
  "arguments": {
    "fix": true,
    "select": ["E", "F", "I", "UP"]
  }
}
```

### Use Case 2: Check Formatting
```bash
{
  "tool": "python_format",
  "arguments": {
    "check": true
  }
}
```

### Use Case 3: Lint Specific Files
```bash
{
  "tool": "python_lint",
  "arguments": {
    "files": ["src/api.py", "src/models.py"],
    "outputFormat": "json"
  }
}
```

## References

**Internal:**
- Lint tools: `src/tools/lint-tools.ts`
- Go format: `src/tools/go-tools.ts` (goFmt)

**External:**
- ruff documentation: https://docs.astral.sh/ruff/
- ruff rules: https://docs.astral.sh/ruff/rules/
- ruff configuration: https://docs.astral.sh/ruff/configuration/

## Success Metrics

- [ ] Linting runs 10x+ faster than flake8
- [ ] Formatting runs 10x+ faster than black
- [ ] Auto-fix resolves 80%+ of common issues
- [ ] All quality gates passed

--- a/src/__tests__/tools/python-tools.test.ts
+++ b/src/__tests__/tools/python-tools.test.ts
@@ -445,4 +445,31 @@ describe("PythonTools", () => {
       expect(result).toContain("--pre");
     });
   });
+
+  describe("Phase 1 enhancements: uv command building with mode support", () => {
+    it("should build uv command with system and editable flags", () => {
+      const pythonTools = new PythonTools();
+      const buildCommand = (pythonTools as any).buildUvCommand.bind(
+        pythonTools,
+      );
+
+      const resultSystem = buildCommand({ mode: "install", system: true });
+      expect(resultSystem).toContain("--system");
+
+      const resultEditable = buildCommand({ mode: "install", editable: true });
+      expect(resultEditable).toContain("-e");
+    });
+
+    it("should build uv install command with upgrade flag for update mode", () => {
+      const pythonTools = new PythonTools();
+      const buildCommand = (pythonTools as any).buildUvCommand.bind(
+        pythonTools,
+      );
+
+      const result = buildCommand({ mode: "update" });
+      expect(result).toContain("pip");
+      expect(result).toContain("install");
+      expect(result).toContain("--upgrade");
+    });
+  });
 });

--- a/src/index.ts
+++ b/src/index.ts
@@ -3066,7 +3066,7 @@ class MCPDevToolsServer {
           }
 
           case "python_format": {
-            const validatedArgs = PythonTools.validateLintArgs(args);
+            const validatedArgs = PythonTools.validateFormatArgs(args);
             const result = await this.pythonTools.pythonFormat(validatedArgs);
             return {
               content: [

--- a/src/tools/python-tools.ts
+++ b/src/tools/python-tools.ts
@@ -36,6 +36,19 @@ const PythonTestArgsSchema = z.object({
     .string()
     .optional()
     .describe("Run tests matching given mark expression (-m)"),
+  parallel: z
+    .boolean()
+    .optional()
+    .describe("Run tests in parallel with pytest-xdist"),
+  maxWorkers: z
+    .number()
+    .optional()
+    .describe("Number of parallel workers (default: auto)"),
+  failFast: z
+    .boolean()
+    .optional()
+    .describe("Stop on first failure (-x)"),
+  junitXml: z.string().optional().describe("Output JUnit XML to file"),
   args: z.array(z.string()).optional().describe("Additional arguments"),
   timeout: z.number().optional().describe("Command timeout in milliseconds"),
 });
@@ -48,12 +61,56 @@ const PythonLintArgsSchema = z.object({
     .array(z.string())
     .optional()
     .describe("Specific files to lint/format"),
+  select: z
+    .array(z.string())
+    .optional()
+    .describe('Rule codes to enable (e.g., ["E", "F", "I"])'),
+  ignore: z.array(z.string()).optional().describe("Rule codes to ignore"),
+  outputFormat: z
+    .enum(["text", "json", "github"])
+    .optional()
+    .describe("Output format"),
+  showFixes: z.boolean().optional().describe("Show available fixes"),
+  args: z.array(z.string()).optional().describe("Additional arguments"),
+  timeout: z.number().optional().describe("Command timeout in milliseconds"),
+});
+
+const PythonFormatArgsSchema = z.object({
+  directory: z.string().optional().describe("Working directory"),
+  check: z
+    .boolean()
+    .optional()
+    .describe("Check without modifying"),
+  files: z.array(z.string()).optional().describe("Specific files to format"),
+  lineLength: z.number().optional().describe("Max line length"),
+  preview: z
+    .boolean()
+    .optional()
+    .describe("Enable preview style"),
   args: z.array(z.string()).optional().describe("Additional arguments"),
   timeout: z.number().optional().describe("Command timeout in milliseconds"),
 });
 
 const PythonTypeCheckArgsSchema = z.object({
   directory: z.string().optional().describe("Working directory"),
+  files: z.array(z.string()).optional().describe("Specific files to check"),
+  level: z
+    .enum(["basic", "standard", "strict"])
+    .optional()
+    .describe("Type checking strictness level"),
+  outputFormat: z
+    .enum(["text", "json"])
+    .optional()
+    .describe("Output format"),
+  createStubs: z
+    .boolean()
+    .optional()
+    .describe("Create type stubs for untyped libraries"),
+  pythonVersion: z
+    .string()
+    .optional()
+    .describe('Target Python version (e.g., "3.11")'),
+  showStats: z.boolean().optional().describe("Show statistics"),
   watch: z.boolean().optional().describe("Watch mode"),
   verbose: z.boolean().optional().describe("Enable verbose output"),
   args: z.array(z.string()).optional().describe("Additional arguments"),
@@ -62,6 +119,14 @@ const PythonTypeCheckArgsSchema = z.object({
 
 const PythonInstallDepsArgsSchema = z.object({
   directory: z.string().optional().describe("Working directory"),
+  mode: z
+    .enum(["install", "sync", "update", "add", "remove"])
+    .optional()
+    .describe("Installation mode"),
+  packages: z
+    .array(z.string())
+    .optional()
+    .describe("Packages to add/remove (for add/remove modes)"),
   packageManager: z
     .enum(["auto", "uv", "poetry", "pipenv", "pip"])
     .optional()
@@ -70,6 +135,19 @@ const PythonInstallDepsArgsSchema = z.object({
     .boolean()
     .optional()
     .describe("Install development dependencies too"),
+  upgrade: z.boolean().optional().describe("Upgrade packages"),
+  prerelease: z
+    .enum(["allow", "if-necessary", "disallow"])
+    .optional()
+    .describe("Pre-release handling (uv only)"),
+  system: z
+    .boolean()
+    .optional()
+    .describe("Install to system Python (uv only)"),
+  editable: z
+    .boolean()
+    .optional()
+    .describe("Install in editable mode"),
   args: z.array(z.string()).optional().describe("Additional arguments"),
   timeout: z.number().optional().describe("Command timeout in milliseconds"),
 });
@@ -89,6 +167,7 @@ const PythonVersionArgsSchema = z.object({
 export type PythonProjectInfoArgs = z.infer<typeof PythonProjectInfoSchema>;
 export type PythonTestArgs = z.infer<typeof PythonTestArgsSchema>;
 export type PythonLintArgs = z.infer<typeof PythonLintArgsSchema>;
+export type PythonFormatArgs = z.infer<typeof PythonFormatArgsSchema>;
 export type PythonTypeCheckArgs = z.infer<typeof PythonTypeCheckArgsSchema>;
 export type PythonInstallDepsArgs = z.infer<typeof PythonInstallDepsArgsSchema>;
 export type PythonVersionArgs = z.infer<typeof PythonVersionArgsSchema>;
@@ -107,11 +186,20 @@ export interface PythonToolResult {
   suggestions?: string[];
 }
 
+export interface PythonVersionInfo {
+  current: string;
+  required?: string;
+  isEOL: boolean;
+  recommendation: string | null;
+  upgradeReason?: string;
+}
+
 export interface PythonProjectInfo {
   hasPyprojectToml: boolean;
   hasSetupPy: boolean;
   hasRequirementsTxt: boolean;
   pythonVersion?: string;
+  pythonVersionInfo?: PythonVersionInfo;
   packageManager?: string;
   projectName?: string;
   projectVersion?: string;
@@ -229,6 +317,110 @@ export class PythonTools {
   }
 
   /**
+   * Check if Python version is End of Life
+   */
+  private isPythonEOL(major: number, minor: number): boolean {
+    const eolVersions = [
+      { major: 3, minor: 7, eolDate: "2023-06-27" },
+      { major: 3, minor: 8, eolDate: "2024-10-14" },
+      { major: 3, minor: 9, eolDate: "2025-10-05" },
+    ];
+
+    return eolVersions.some(
+      (v) =>
+        v.major === major &&
+        v.minor === minor &&
+        new Date() > new Date(v.eolDate),
+    );
+  }
+
+  /**
+   * Get required Python version from pyproject.toml
+   */
+  private async getRequiredPythonVersion(
+    directory: string,
+  ): Promise<string | undefined> {
+    const pyprojectPath = path.join(directory, "pyproject.toml");
+    try {
+      await fs.access(pyprojectPath);
+      const content = await fs.readFile(pyprojectPath, "utf-8");
+      const match = content.match(/requires-python\s*=\s*"([^"]+)"/);
+      return match ? match[1] : undefined;
+    } catch {
+      return undefined;
+    }
+  }
+
+  /**
+   * Check Python version and generate PythonVersionInfo
+   */
+  private async checkPythonVersion(
+    directory: string,
+  ): Promise<PythonVersionInfo | undefined> {
+    try {
+      const pythonExecutable = await this.detectPythonExecutable();
+      if (!pythonExecutable) {
+        return undefined;
+      }
+
+      const versionResult = await this.executor.execute(pythonExecutable, {
+        cwd: directory,
+        args: ["--version"],
+        timeout: PythonTools.TIMEOUT_VERSION_CHECK,
+      });
+
+      if (!versionResult.success) {
+        return undefined;
+      }
+
+      const versionMatch = versionResult.stdout.match(/Python (\d+\.\d+\.\d+)/);
+      const current = versionMatch ? versionMatch[1] : "unknown";
+
+      if (current === "unknown") {
+        return {
+          current,
+          isEOL: false,
+          recommendation: null,
+        };
+      }
+
+      // Parse version
+      const [major, minor] = current.split(".").map(Number);
+
+      // Check if EOL
+      const isEOL = this.isPythonEOL(major, minor);
+
+      // Get required version from pyproject.toml
+      const required = await this.getRequiredPythonVersion(directory);
+
+      // Generate recommendation (using existing method)
+      const recommendation = this.shouldRecommendUpgrade(current)
+        ? this.getUpgradeRecommendation(current)
+        : null;
+
+      // Get upgrade reason
+      const upgradeReason =
+        major !== 3
+          ? undefined
+          : minor <= 9
+            ? "Security updates discontinued, performance improvements available"
+            : minor === 10
+              ? "Approaching end of life"
+              : undefined;
+
+      return {
+        current,
+        required,
+        isEOL,
+        recommendation,
+        upgradeReason,
+      };
+    } catch {
+      return undefined;
+    }
+  }
+
+  /**
    * Get Python project information
    * Detects Python version, package manager, dependencies, and project structure
    */
@@ -237,10 +429,10 @@ export class PythonTools {
   ): Promise<PythonProjectInfo> {
     const dir = args.directory || this.projectRoot;
 
-    // Try cache first
+    // Try cache first (uses projectDetection namespace: 60s TTL)
     const cacheKey = this.buildPythonCacheKey("project-info", { directory: dir });
     const cached = this.cacheManager.get<PythonProjectInfo>(
-      "pythonTools",
+      "projectDetection",
       cacheKey,
     );
     if (cached) {
@@ -357,37 +549,18 @@ export class PythonTools {
         }
       }
 
-      // Detect Python version using the cached detection method
-      let pythonVersion: string | undefined;
+      // Detect Python version and generate version info
       const pythonExecutable = await this.detectPythonExecutable();
+      const versionInfo = await this.checkPythonVersion(dir);
 
-      if (pythonExecutable) {
-        const versionResult = await this.executor.execute(pythonExecutable, {
-          cwd: dir,
-          args: ["--version"],
-          timeout: PythonTools.TIMEOUT_VERSION_CHECK,
-        });
-
-        if (versionResult.success) {
-          // Output format: "Python 3.9.18"
-          const match = versionResult.stdout.match(/Python\s+([\d.]+)/);
-          if (match) {
-            pythonVersion = match[1];
-          }
-        }
-      }
-
-      if (pythonVersion && pythonExecutable) {
-        info.pythonVersion = pythonVersion;
+      if (versionInfo && pythonExecutable) {
+        info.pythonVersion = versionInfo.current;
+        info.pythonVersionInfo = versionInfo;
         info.pythonExecutable = pythonExecutable;
-        info.supportsModernFeatures = !this.shouldRecommendUpgrade(
-          pythonVersion,
-        );
+        info.supportsModernFeatures = !versionInfo.isEOL && !versionInfo.recommendation;
 
-        if (this.shouldRecommendUpgrade(pythonVersion)) {
-          info.upgradeRecommendation = this.getUpgradeRecommendation(
-            pythonVersion,
-          );
+        if (versionInfo.recommendation) {
+          info.upgradeRecommendation = versionInfo.recommendation;
         }
       }
 
@@ -470,8 +643,8 @@ export class PythonTools {
         }
       }
 
-      // Cache the result with 5-minute TTL
-      this.cacheManager.set("pythonTools", cacheKey, info);
+      // Cache the result (uses projectDetection namespace: 60s TTL)
+      this.cacheManager.set("projectDetection", cacheKey, info);
 
       return info;
     } catch (error) {
@@ -518,6 +691,19 @@ export class PythonTools {
     const dir = args.directory || this.projectRoot;
     const validated = PythonTestArgsSchema.parse(args);
 
+    // Check cache (uses testResults namespace: 60s TTL)
+    const cacheKey = this.buildPythonCacheKey("test", args);
+    const cached = this.cacheManager.get<PythonToolResult>(
+      "testResults",
+      cacheKey,
+    );
+    if (cached) {
+      return {
+        ...cached,
+        command: cached.command + " (cached)",
+      };
+    }
+
     const startTime = performance.now();
 
     try {
@@ -533,6 +719,21 @@ export class PythonTools {
       // Add verbose flag
       if (validated.verbose) {
         commandArgs.push("-vv");
+      }
+
+      // Add parallel execution
+      if (validated.parallel) {
+        commandArgs.push("-n", validated.maxWorkers?.toString() || "auto");
+      }
+
+      // Add fail fast
+      if (validated.failFast) {
+        commandArgs.push("-x");
+      }
+
+      // Add JUnit XML output
+      if (validated.junitXml) {
+        commandArgs.push(`--junit-xml=${validated.junitXml}`);
       }
 
       // Add markers (e.g., -m "unit" to run tests marked as unit tests)
@@ -565,7 +766,14 @@ export class PythonTools {
       });
 
       const duration = performance.now() - startTime;
-      return this.processPythonResult(result, "pytest", duration);
+      const pythonResult = this.processPythonResult(result, "pytest", duration);
+
+      // Cache successful test results (uses testResults namespace: 60s TTL)
+      if (pythonResult.success) {
+        this.cacheManager.set("testResults", cacheKey, pythonResult);
+      }
+
+      return pythonResult;
     } catch (error) {
       const duration = performance.now() - startTime;
       const errorMsg = error instanceof Error ? error.message : String(error);
@@ -612,6 +820,26 @@ export class PythonTools {
       // Add fix flag (auto-fix issues)
       if (validated.fix) {
         commandArgs.push("--fix");
+      }
+
+      // Add output format
+      if (validated.outputFormat && validated.outputFormat !== "text") {
+        commandArgs.push("--output-format", validated.outputFormat);
+      }
+
+      // Add rule selection
+      if (validated.select && validated.select.length > 0) {
+        commandArgs.push("--select", validated.select.join(","));
+      }
+
+      // Add rule ignores
+      if (validated.ignore && validated.ignore.length > 0) {
+        commandArgs.push("--ignore", validated.ignore.join(","));
+      }
+
+      // Add show fixes
+      if (validated.showFixes) {
+        commandArgs.push("--show-fixes");
       }
 
       // Add files to lint, or current directory
@@ -669,8 +897,8 @@ export class PythonTools {
    * @param args - Formatting configuration with directory, files, check, preview options
    * @returns Result with formatting output and suggestions
    */
-  async pythonFormat(args: PythonLintArgs): Promise<PythonToolResult> {
-    const validated = PythonLintArgsSchema.parse(args);
+  async pythonFormat(args: PythonFormatArgs): Promise<PythonToolResult> {
+    const validated = PythonFormatArgsSchema.parse(args);
     const directory = validated.directory || this.projectRoot;
     const startTime = performance.now();
 
@@ -680,6 +908,16 @@ export class PythonTools {
       // Add check flag (verify without modifying)
       if (validated.check) {
         commandArgs.push("--check");
+      }
+
+      // Add line length
+      if (validated.lineLength) {
+        commandArgs.push("--line-length", validated.lineLength.toString());
+      }
+
+      // Add preview mode
+      if (validated.preview) {
+        commandArgs.push("--preview");
       }
 
       // Add files to format, or current directory
@@ -756,6 +994,34 @@ export class PythonTools {
       // Build pyright command
       const commandArgs: string[] = [];
 
+      // Add output format
+      if (validated.outputFormat === "json") {
+        commandArgs.push("--outputjson");
+      }
+
+      // Add strictness level (default: standard)
+      const level = validated.level || "standard";
+      if (level === "strict") {
+        commandArgs.push("--level", "error");
+      } else if (level === "basic") {
+        commandArgs.push("--level", "warning");
+      }
+
+      // Add Python version
+      if (validated.pythonVersion) {
+        commandArgs.push("--pythonversion", validated.pythonVersion);
+      }
+
+      // Add stats
+      if (validated.showStats) {
+        commandArgs.push("--stats");
+      }
+
+      // Add create stubs
+      if (validated.createStubs) {
+        commandArgs.push("--createstub");
+      }
+
       // Add watch mode
       if (validated.watch) {
         commandArgs.push("--watch");
@@ -764,6 +1030,12 @@ export class PythonTools {
       // Add verbose output
       if (validated.verbose) {
         commandArgs.push("--verbose");
+      }
+
+      // Add specific files to check
+      if (validated.files && validated.files.length > 0) {
+        this.validateFilePaths(validated.files);
+        commandArgs.push(...validated.files);
       }
 
       // Add any additional arguments
@@ -1050,26 +1322,75 @@ export class PythonTools {
   }
 
   /**
-   * Build uv install command
-   * uv pip install [--all-extras] [--upgrade] [--prerelease {allow,if-necessary,disallow}]
+   * Build uv command based on mode
+   * Supports: install, sync, update, add, remove
    */
   private buildUvCommand(args: PythonInstallDepsArgs): string[] {
-    const commandArgs = ["pip", "install"];
+    const mode = args.mode || "install";  // Default: install
+    const commandArgs = ["pip"];
+
+    switch (mode) {
+      case "install":
+        commandArgs.push("install");
+        break;
+      case "sync":
+        commandArgs.push("sync");
+        if (args.packages && args.packages.length > 0) {
+          commandArgs.push(...args.packages);
+        } else {
+          commandArgs.push("requirements.txt");
+        }
+        return commandArgs;
+      case "update":
+        commandArgs.push("install", "--upgrade");
+        break;
+      case "add":
+        commandArgs.push("install");
+        break;
+      case "remove":
+        commandArgs.push("uninstall");
+        if (args.packages && args.packages.length > 0) {
+          commandArgs.push(...args.packages);
+        }
+        return commandArgs;
+    }
+
+    // Add upgrade flag
+    if (args.upgrade && mode !== "update") {
+      commandArgs.push("--upgrade");
+    }
+
+    // Add prerelease handling
+    if (args.prerelease) {
+      commandArgs.push("--prerelease", args.prerelease);
+    }
+
+    // Add system flag
+    if (args.system) {
+      commandArgs.push("--system");
+    }
+
+    // Add editable flag
+    if (args.editable) {
+      commandArgs.push("-e");
+    }
 
     // Install dev dependencies (extras)
-    if (args.dev) {
+    if (args.dev && mode === "install") {
       commandArgs.push("--all-extras");
+    }
+
+    // Add packages or install from project
+    if (args.packages && args.packages.length > 0) {
+      commandArgs.push(...args.packages);
+    } else if (mode === "install" || mode === "update") {
+      // Install from pyproject.toml or requirements.txt
+      commandArgs.push(".");
     }
 
     // Add any additional arguments
     if (args.args && args.args.length > 0) {
       commandArgs.push(...args.args);
-    }
-
-    // If no specific arguments, install from requirements.txt or pyproject.toml
-    if (!args.args || args.args.length === 0) {
-      // Try to install from pyproject.toml (modern Python projects)
-      commandArgs.push(".");
     }
 
     return commandArgs;
@@ -1120,40 +1441,70 @@ export class PythonTools {
   }
 
   /**
-   * Build pip install command
-   * pip install [-r requirements.txt] [--upgrade] [--pre]
+   * Build pip command based on mode
+   * Supports: install, sync, update, add, remove
    */
   private buildPipCommand(args: PythonInstallDepsArgs): string[] {
-    const commandArgs = ["install"];
+    const mode = args.mode || "install";  // Default: install
     const dir = args.directory || this.projectRoot;
+    const commandArgs = [];
 
-    // Check for requirements.txt before adding it
-    const reqPath = path.join(dir, "requirements.txt");
-    const reqDevPath = path.join(dir, "requirements-dev.txt");
+    switch (mode) {
+      case "install":
+      case "sync":
+        commandArgs.push("install");
+        break;
+      case "update":
+        commandArgs.push("install", "--upgrade");
+        break;
+      case "add":
+        commandArgs.push("install");
+        break;
+      case "remove":
+        commandArgs.push("uninstall", "-y");
+        if (args.packages && args.packages.length > 0) {
+          commandArgs.push(...args.packages);
+        }
+        return commandArgs;
+    }
 
-    try {
-      // Check if requirements.txt exists
-      const hasRequirements = existsSync(reqPath);
-      if (hasRequirements) {
+    // Add upgrade flag
+    if (args.upgrade && mode !== "update") {
+      commandArgs.push("--upgrade");
+    }
+
+    // Add editable flag
+    if (args.editable) {
+      commandArgs.push("-e");
+    }
+
+    // Add packages or install from requirements
+    if (args.packages && args.packages.length > 0) {
+      commandArgs.push(...args.packages);
+    } else if (mode === "install" || mode === "sync" || mode === "update") {
+      // Check for requirements.txt
+      const reqPath = path.join(dir, "requirements.txt");
+      const reqDevPath = path.join(dir, "requirements-dev.txt");
+
+      try {
+        const hasRequirements = existsSync(reqPath);
+        if (hasRequirements) {
+          commandArgs.push("-r", "requirements.txt");
+        }
+
+        // Handle dev dependencies
+        if (args.dev) {
+          const hasDevRequirements = existsSync(reqDevPath);
+          if (hasDevRequirements) {
+            commandArgs.push("-r", "requirements-dev.txt");
+          }
+        }
+      } catch {
+        // If file system check fails, proceed with defaults
         commandArgs.push("-r", "requirements.txt");
-      } else {
-        // If no requirements.txt, still continue (pip will handle it)
-        // but log a warning suggestion
-      }
-
-      // Handle dev dependencies
-      if (args.dev) {
-        const hasDevRequirements = existsSync(reqDevPath);
-        if (hasDevRequirements) {
+        if (args.dev) {
           commandArgs.push("-r", "requirements-dev.txt");
         }
-      }
-    } catch {
-      // If file system check fails, proceed with defaults
-      // pip will provide appropriate error messages
-      commandArgs.push("-r", "requirements.txt");
-      if (args.dev) {
-        commandArgs.push("-r", "requirements-dev.txt");
       }
     }
 
@@ -1304,6 +1655,10 @@ Installation:
 
   static validateLintArgs(args: unknown): PythonLintArgs {
     return PythonLintArgsSchema.parse(args);
+  }
+
+  static validateFormatArgs(args: unknown): PythonFormatArgs {
+    return PythonFormatArgsSchema.parse(args);
   }
 
   static validateTypeCheckArgs(args: unknown): PythonTypeCheckArgs {

